### PR TITLE
=act #16969 Props.create must not check STATIC flag

### PR DESCRIPTION
**Pull request by patriknw**
I reproduced the issue with Eclipse. After starting at this for quite some time I came up with this, which passes the test.

It was originally written by @ktoso  in https://github.com/akka/akka/pull/17412

### CODE REVIEW

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 1]
```diff
@@ -59,6 +60,19 @@ private[akka] trait AbstractProps {
...
+      try { clazz.getDeclaredConstructor(); true }
+      catch { case ex: NoSuchMethodException ⇒ false }
+
+    def checkConstructors = {
+      val constructorsLength = clazz.getConstructors.length
+      constructorsLength > 0 || (constructorsLength == 0 && hasDeclaredConstructorWithEmptyParams)
+    }
+
+    (clazz.getEnclosingClass eq null) || checkConstructors
+  }
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: I changed this part, compare with https://github.com/akka/akka/pull/17412/files#r30890616

**Comment by patriknw** [comment]
Refs #16969

**Comment by drewhk** [comment]
LGTM

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 1 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: Hmmm, I think looks good hmm

**Comment by ktoso** [comment]
interesting, it breaks the lambda example in https://github.com/akka/akka/blob/master/akka-docs/rst/java/code/docs/actor/ActorDocTest.java#L126
Not quite sure if that's right... Perhaps it is right actually hm

**Comment by patriknw** [comment]
@ktoso it was not the lamda that was the problem. It was that we actually support this case (and type is not erased):

```scala
    public static Props propsUsingCreatorWithoutClass(final int magicNumber) {
      return Props.create(new Creator<UntypedTestActor>() {
        private static final long serialVersionUID = 1L;

        @Override
        public UntypedTestActor create() throws Exception {
          return new UntypedTestActor(magicNumber);
        }
      });
    }
```

I have added support for that also: https://github.com/akka/akka/pull/20488/commits/0822ba89138b80da931593315c080a56ea72194f. I find it very difficult to know if we cover all cases.

**Comment by ktoso** [comment]
Yeah it's very hard to reason about all the cases here. Reviewing the latest commit now

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 2]
```diff
@@ -60,23 +59,57 @@ private[akka] trait AbstractProps {
...
+    val enclosingClass = clazz.getEnclosingClass
+
+    def hasDeclaredConstructorWithEmptyParams(declaredConstructors: Array[Constructor[_]]): Boolean = {
+      @tailrec def loop(i: Int): Boolean = {
+        if (i == declaredConstructors.length) false
+        else {
+          if (declaredConstructors(i).getParameterCount == 0)
+            true
+          else
+            loop(i + 1) // recur
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: Could be `declaredConstructors(I).getParameterCount == 0 || loop(I + 1)`

Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: yes, good

**Comment by ktoso** [comment]
LGTM I think.

**Pull request merged**